## SQL Practice

> Authors: Caroline Schmitt & Katie Sylvia


In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

**Data**: The data used in this lab is the famous (but fake) Northwind database. It contains various tables of transactions. The schema for the Northwind database can be found [here](./assets/Northwind_ERD.png).

**Important**: The table and label names in the SQLite database are **slightly different than the ERD**. In particular, many of the tables names are singular where the ERD lists them as plural, e.g. the `Customers` table in the ERD is actually called `Customer` in the SQLite database. The instructions should provide the correct table names. The field names themselves should be correct.

You'll be using Python's `sqlite3` library this time. `sqlite3` comes with Python, so you shouldn't have to install anything. You will once again be passing queries into `pd.read_sql_query()` as strings. If you try to close the notebook and get an alert that there is still a connection to the database, you can run `con.close()` to close the connection. (Alternatively, it should be safe to force exit.)

In [1]:
import pandas as pd
from pandas.io import sql

import sqlite3

### Prompt 1

How many rows are there in the `Customer` table?

In [2]:
con = sqlite3.connect('data/Northwind_small.sqlite')

# modify this SQL query to answer the question
sql = """
SELECT COUNT(*)
FROM Customer;
"""

pd.read_sql_query(sql, con)

COUNT(*)
0        91

**Prompt 1 Answer**: 

There are 91 rows of data in the Customer table!

### Prompt 2

If you sort the customers in alphabetical order by name, which customer is fourth?

(You can answer this by retrieving the first four customers; you _do not_ need to retrieve only the 4th customer when customers are sorted alphabetically.)

In [3]:
con = sqlite3.connect('data/Northwind_small.sqlite')

sql = """
SELECT ContactName
FROM Customer
ORDER BY ContactName 
LIMIT 4
"""
#order by is alphabetical order
#DESC - reverse alphabetical order
pd.read_sql_query(sql, con)

ContactName
0   Alejandra Camino
1    Alexander Feuer
2       Ana Trujillo
3  Anabela Domingues

**Prompt 2 Answer**: 



The customer that is 4th when sorted in alphabetical order is Anabela Domingues.

### Prompt 3

Among those customers who live in Germany, if you sort the customers in alphabetical order by name, which customer is fourth?

In [12]:
# Use a SQL query to find the answer to the above prompt.
con = sqlite3.connect('data/Northwind_small.sqlite')

sql = """
SELECT ContactName
FROM Customer
WHERE Country = 'Germany'
ORDER BY ContactName
LIMIT 4
"""

pd.read_sql_query(sql, con)

ContactName
0      Alexander Feuer
1           Hanna Moos
2  Henriette Pfalzheim
3          Horst Kloss

**Prompt 3 Answer**: 

The customer in Germany that appears 4th in alphabetical order is Horst Kloss.

### Prompt 4

From the `Employee` table, what are the first names of the individuals who have a birthdate in the 1960s?

In [18]:
# Use a SQL query to find the answer to the above prompt.
con = sqlite3.connect('data/Northwind_small.sqlite')

sql = """

SELECT FirstName
FROM Employee
WHERE BirthDate BETWEEN '1960-01-01' AND '1969-12-31'
"""

pd.read_sql_query(sql, con)

FirstName
0  Margaret

**Prompt 4 Answer**: 

Only Margaret has a birthdate in the 1960s.

### Prompt 5

In `OrderDetail` table, what is the total number of units sold for products 2, 3, and 4? 

You should give us three numbers - one for each `ProductID`.

> You can solve this with either three separate queries, or with just one query.

In [29]:
# Use a SQL query to find the answer to the above prompt.
con = sqlite3.connect('data/Northwind_small.sqlite')

sql = """
SELECT ProductId, SUM(Quantity)
FROM OrderDetail
WHERE ProductId IN (2,3,4)
GROUP BY ProductId 
"""
#Group By is used when doing aggregate function - group data together
pd.read_sql_query(sql, con)

ProductId  SUM(Quantity)
0          2           1057
1          3            328
2          4            453

**Prompt 5 Answer**:

The total number of units sold for product 2 is 1057, for product 3 is 328, and for product 4 is 453.

### Prompt 6

From the `OrderDetail` table, calculate the revenue per product. In your SQL output, rename the revenue column `Revenue` and sort your results from largest revenue per product to smallest revenue per product. Which `ProductID` has the fifth-highest revenue, and what is that revenue rounded to the nearest dollar?

Note that revenue for a given product should be equal to the total unit price of that product times the quantity. You do not need to worry about the discount here. If you haven't multiplied columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [31]:
# Use a SQL query to find the answer to the above prompt.
con = sqlite3.connect('data/Northwind_small.sqlite')

sql = """
SELECT ProductId, SUM(UnitPrice * Quantity) AS Revenue
FROM OrderDetail
GROUP BY ProductId
ORDER BY Revenue
DESC
LIMIT 5
"""

pd.read_sql_query(sql, con)

ProductId   Revenue
0         38  149984.2
1         29   87736.4
2         59   76296.0
3         60   50286.0
4         62   49827.9

ProductID 62 has the 5th highest revenue: $49,828.

**Prompt 6 Answer**:

### Prompt 7

From the `Product` table, show the `ProductName`, `UnitsInStock`, and `UnitsOnOrder` values for those products who have more units on order than units in stock. Among these results, which product has the most units on order?

In [27]:
# Use a SQL query to find the answer to the above prompt.
con = sqlite3.connect('data/Northwind_small.sqlite')

sql = """
SELECT
    ProductName, UnitsInStock, UnitsOnOrder
FROM Product
WHERE UnitsOnOrder > UnitsInStock
ORDER BY UnitsOnOrder
DESC
LIMIT 1
"""

pd.read_sql_query(sql, con)

ProductName  UnitsInStock  UnitsOnOrder
0  Louisiana Hot Spiced Okra             4           100

**Prompt 7 Answer**:

Lousiana Hot Spiced Okra has the most units on order.

### Prompt 8

From the `OrderDetail` table, show the products that have an average discount of at least 8%. Among these results, which average discount is closest to 8%?

In [51]:
# Use a SQL query to find the answer to the above prompt.
con = sqlite3.connect('data/Northwind_small.sqlite')

sql = """
SELECT ProductId, AVG(Discount) AS avg_discount
FROM OrderDetail
GROUP BY ProductId
HAVING avg_discount >= .08
ORDER BY avg_discount
ASC
"""
#having allows to filter table after using group by

pd.read_sql_query(sql, con)

ProductId  avg_discount
0         30      0.081250
1         25      0.086111
2          9      0.100000
3         58      0.100000
4          2      0.102273
5         48      0.108333

**Prompt 8 Answer**:

ProductId 30 has the closest average discount of 8%.

### Prompt 9

Return the `Orders` table, but also include the first name, last name, and country of the employee assigned to each order (you'll need the `Employee` table for this.) What is the first and last name of the employee assigned to OrderID 11077?

In [86]:
con = sqlite3.connect('data/Northwind_small.sqlite')

sql = """
Select FirstName, LastName, Country, o.Id
FROM Orders AS o
LEFT JOIN Employee AS e
ON e.Id = o.EmployeeID
WHERE o.Id = 11077
"""

pd.read_sql_query(sql, con)

FirstName LastName Country     Id
0     Nancy  Davolio     USA  11077

Nancy Davolio is the first and last name of the employee assigned to OrderId 11077.

**Prompt 9 Answer**:

### Prompt 10

Find the average amount of freight across all orders, broken out by country of *employee*. (**NOT `ShipCountry`**.) How much higher is the average freight from UK employees than the average freight from US employees?

> For a bonus, round the freight to the nearest whole number and rename the column `Average Freight`.

In [89]:
con = sqlite3.connect('data/Northwind_small.sqlite')

sql = """
SELECT e.Country, AVG(Freight)
FROM Orders AS o
LEFT JOIN Employee AS e
ON o.EmployeeId = e.Id
GROUP BY e.Country;
"""

pd.read_sql_query(sql, con)

Country  AVG(Freight)
0      UK     78.977143
1     USA     77.973284

**Prompt 10 Answer**:

---

Once you're done, run the following cell to close your connection to the database:

The difference between the two average is 1.003859.

In [ ]:
con.close()